src/widget/progress_bar.rs

```diff
@@ -14,67 +14,69 @@
-use crate::{BoxConstraints, HandlerCtx, Id, LayoutCtx, LayoutResult, PaintCtx, Ui};
-
-use crate::kurbo::{Rect, Size};
+use crate::kurbo::{Point, Size};
 use crate::piet::{Color, FillRule, RenderContext};
+use crate::{
+    Action, BaseState, BoxConstraints, Env, Event, EventCtx, LayoutCtx, PaintCtx,
+    UpdateCtx, Widget,
+};
 
-const BOX_HEIGHT: f64 = 24.;
 const BACKGROUND_COLOR: Color = Color::rgb24(0x55_55_55);
 const BAR_COLOR: Color = Color::rgb24(0xf0_f0_ea);
 
+#[derive(Debug, Clone, Default)]
 pub struct ProgressBar {
-    value: f64,
 }
 
-impl ProgressBar {
-    pub fn new(initial_value: f64) -> ProgressBar {
-        ProgressBar {
-            value: initial_value,
-        }
-    }
-    pub fn ui(self, ctx: &mut Ui) -> Id {
-        ctx.add(self, &[])
-    }
-}
+impl Widget<f64> for ProgressBar {
+    fn paint(
+        &mut self,
+        paint_ctx: &mut PaintCtx, 
+        base_state: &BaseState,
+        data: &f64,
+        _env: &Env,
+    ) {
+        let clamped = data.max(0.0).min(1.0);
+        let rect = base_state.layout_rect.with_origin(Point::ORIGIN);
 
-impl Widget for ProgressBar {
-    fn paint(&mut self, paint_ctx: &mut PaintCtx, geom: &Rect) {
         //Paint the background
         let brush = paint_ctx.render_ctx.solid_brush(BACKGROUND_COLOR);
-
-        paint_ctx.render_ctx.fill(geom, &brush, FillRule::NonZero);
+        paint_ctx.render_ctx.fill(rect, &brush, FillRule::NonZero);
 
         //Paint the bar
         let brush = paint_ctx.render_ctx.solid_brush(BAR_COLOR);
-
-        let calculated_bar_width = self.value * geom.width();
-
-        let rect = geom.with_size(Size::new(calculated_bar_width, geom.height()));
+        let calculated_bar_width = clamped * rect.width();
+        let rect = rect.with_size(Size::new(calculated_bar_width, rect.height()));
         paint_ctx.render_ctx.fill(rect, &brush, FillRule::NonZero);
     }
 
     fn layout(
         &mut self,
+        _layout_ctx: &mut LayoutCtx,
         bc: &BoxConstraints,
-        _children: &[Id],
-        _size: Option<Size>,
-        _ctx: &mut LayoutCtx,
-    ) -> LayoutResult {
-        LayoutResult::Size(bc.constrain((bc.max.width, BOX_HEIGHT)))
+        _data: &f64,
+        _env: &Env
+    ) -> Size {
+        bc.constrain((bc.max.width, bc.max.height))
     }
 
-    fn poke(&mut self, payload: &mut dyn Any, ctx: &mut HandlerCtx) -> bool {
-        if let Some(value) = payload.downcast_ref::<f64>() {
-            self.value = *value;
-            ctx.invalidate();
-            true
-        } else {
-            println!("downcast failed");
-            false
-        }
+    fn event(
+        &mut self,
+        _event: &Event,
+        _ctx: &mut EventCtx,
+        _data: &mut f64,
+        _env: &Env,
+    ) -> Option<Action> {
+        None
     }
-}
+
+    fn update(
+        &mut self,
+        ctx: &mut UpdateCtx,
+        _old_data: Option<&f64>,
+        _data: &f64,
+        _env: &Env,
+    ) {
+        ctx.invalidate();
```

Comment:
I think this is unnecessary. `update` is currently poorly described and I'm not totally clear what its role is supposed to be, but this should work without it.